### PR TITLE
Linear decomposition of multicontrolled SU(2) gates

### PR DESCRIPTION
**Pull request by israelferrazaraujo**

### CODE REVIEW

File: qclib/gates/mcg.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+def _check_su2(matrix):
+    return isclose(np.linalg.det(matrix), 1.0)
+
+def _check_real_diag(matrix):
+    is_primary_diag_real = matrix[0,0].imag == 0.0 and matrix[1,1].imag == 0.0
+    is_secondary_diag_real = matrix[0,1].imag == 0.0 and matrix[1,0].imag == 0.0
+    return is_primary_diag_real or is_secondary_diag_real
+
+
+def linear_mcg_su2(
```
Comment (adjs) [inline qclib/gates/mcg.py:73]: This function should raise an error if unitary is not a SU(2) gate with one real diagonal.

File: qclib/gates/mcg.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    """
+    Multicontrolled gate decomposition with linear cost.
+    `unitary` must be a SU(2) matrix with at least one real diagonal.
+    """
+    is_secondary_diag_real = unitary[0,1].imag == 0.0 and unitary[1,0].imag == 0.0
+
+    if is_secondary_diag_real:
+        x = unitary[0,1]
+        z = unitary[1,1]
+    else:
```
Comment (adjs) [inline qclib/gates/mcg.py:89]: We should include the case when the su2 gate has no real diagonals.

File: qclib/gates/mcg.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        In-place application of k-controlled X gate with k - 2 dirty ancilla qubits.
+        This decomposition uses the optimization shown in Lemma 9 of
+        https://arxiv.org/abs/1501.06911), which contains at most 8k - 6 `cx`.
+
+        Parameters
+        ----------
+        circuit : quantum circuit where the k-controlled X will be applied
+        controls : k control qubits
+        ancillae : at least k - 2 dirty ancilla qubits
+        target : target qubit of the operation
```
Comment (adjs) [inline qclib/gates/mcg.py:128]: Documentation needs to be updated.

File: qclib/gates/mcg.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+def linear_depth_mcv(
+    self,
+    x,
+    z,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    ctrl_state: str=None
+):
+
```
Comment (adjs) [inline qclib/gates/mcg.py:214]: missing docstring

File: qclib/gates/mcg.py (hunk 1 above)
Comment (israelferrazaraujo) [reply qclib/gates/mcg.py:73]: Ok, I'll include the check.

File: qclib/gates/mcg.py (hunk 2 above)
Comment (israelferrazaraujo) [reply qclib/gates/mcg.py:89]: Yes. Ismael will include a function for this, implementing Barenco's Lemma 7.9.

File: qclib/gates/mcg.py (hunk 3 above)
Comment (israelferrazaraujo) [reply qclib/gates/mcg.py:128]: Ok.

File: qclib/gates/mcg.py (hunk 4 above)
Comment (israelferrazaraujo) [reply qclib/gates/mcg.py:214]: We have to do a full style check before merge.

File: qclib/gates/mcg.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    if len(controls) < 3:
+
+        self.append(c_gate, [target])
+        self.mcx(controls, [target])
+        self.append(b_gate, [target])
+        self.mcx(controls, [target])
+        self.append(a_gate, [target])
+
+    else:
+        ancilla = controls[-1]
```
Comment (israelferrazaraujo) [inline qclib/gates/mcg.py:363]: @IsmaelCesar, could you check that everything is okay here?

Comment (IsmaelCesar) [reply qclib/gates/mcg.py:363]: Everything seems okay.

File: qclib/gates/mcx_gate.py [hunk 6]
```diff
@@ -153,26 +148,27 @@ def linear_mcx(
...
         )
-        self.mcx_v_chain_dirty(
+        mcx_v_chain_dirty(
+            self,
             control_qubits=control_qubits[:k_1],
             target_qubit=ancilla_qubits,
             ancilla_qubits=control_qubits[k_1:k_1 + k_1 - 2]
         )
-        self.mcx_v_chain_dirty(
+        mcx_v_chain_dirty(
```
Comment (adjs) [inline qclib/gates/mcx_gate.py:169]: action_only=True and invert this gate.

File: qclib/gates/mcx_gate.py [hunk 7]
```diff
@@ -153,26 +148,27 @@ def linear_mcx(
         k_1 = int(np.ceil((num_ctrl + 1.) / 2.))
         k_2 = int(np.floor((num_ctrl + 1.) / 2.))
 
-        self.mcx_v_chain_dirty(
+        mcx_v_chain_dirty(
```
Comment (adjs) [inline qclib/gates/mcx_gate.py:151]: action_only=True

Comment (adjs) [reply qclib/gates/mcx_gate.py:151]: isso mesmo, última figura da página 13 e primeira da página 14

**Comment by adjs** [comment]
Closed in favor of PR #123

**Pull request closed without merging**